lol_hub.py

```diff
@@ -0,0 +1,608 @@
+import errno
+import hashlib
+import json
+import os
+import re
+import shutil
+import sys
+import tempfile
+import torch
+import warnings
+import zipfile
+
+from urllib.error import HTTPError
+from urllib.request import urlopen, Request
+from urllib.parse import urlparse  # noqa: F401
+
+try:
+    from tqdm.auto import tqdm  # automatically select proper tqdm submodule if available
+except ImportError:
+    try:
+        from tqdm import tqdm
+    except ImportError:
+        # fake tqdm if it's not installed
+        class tqdm(object):  # type: ignore[no-redef]
+
+            def __init__(self, total=None, disable=False,
+                         unit=None, unit_scale=None, unit_divisor=None):
+                self.total = total
+                self.disable = disable
+                self.n = 0
+                # ignore unit, unit_scale, unit_divisor; they're just for real tqdm
+
+            def update(self, n):
+                if self.disable:
+                    return
+
+                self.n += n
+                if self.total is None:
+                    sys.stderr.write("\r{0:.1f} bytes".format(self.n))
+                else:
+                    sys.stderr.write("\r{0:.1f}%".format(100 * self.n / float(self.total)))
+                sys.stderr.flush()
+
+            def close(self):
+                self.disable = True
+
+            def __enter__(self):
+                return self
+
+            def __exit__(self, exc_type, exc_val, exc_tb):
+                if self.disable:
+                    return
+
+                sys.stderr.write('\n')
+
+# matches bfd8deac from resnet18-bfd8deac.pth
+HASH_REGEX = re.compile(r'-([a-f0-9]*)\.')
+
+ENV_GITHUB_TOKEN = 'GITHUB_TOKEN'
+ENV_TORCH_HOME = 'TORCH_HOME'
+ENV_XDG_CACHE_HOME = 'XDG_CACHE_HOME'
+DEFAULT_CACHE_DIR = '~/.cache'
+VAR_DEPENDENCY = 'dependencies'
+MODULE_HUBCONF = 'hubconf.py'
+READ_DATA_CHUNK = 8192
+_hub_dir = None
+
+
+def _restore_import_cache(func):
+    # call func, and resore the sys.modules dict to what it was prior to the call
+    # This avoids bugs like https://github.com/pytorch/hub/issues/243
+    def inner(*args, **kwargs):
+        before = set(sys.modules.keys())
+        out = func(*args, **kwargs)
+        after = set(sys.modules.keys())
+        for module in after - before:
+            sys.modules.pop(module)
+
+        return out
+
+    return inner
```

Comment:
This decorator and its use in `help()`, `list()`, and `load()` is the only relevant chance of this PR. I need to copy-paste the entire file but the rest is unchanged.